File: cpp/20-valid-parentheses.cpp
class Solution
{
public:
    bool isValid(string s)
    {
        stack<char> stk;
        map<char, char> char_map = {
            {'(', ')'},
            {'{', '}'},
            {'[', ']'}};

        for (auto &ch : s)
        {
            if (char_map.count(ch))
                stk.push(ch);
            else if (stk.empty())
                return false;
            else
            {
                char comp = stk.top();
                stk.pop();
                if (char_map[comp] != ch)
                    return false;
            }
        }

        return stk.empty();
    }
};

File: blind-75/20-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        pair_map = {")": "(",
            "}": "{",
            "]": "["
        }
        stk = []
        for c in s:
            if c in pair_map.values():
                stk.append(c)
            else:
                if len(stk) == 0:
                    return False
                comp = pair_map[c]
                if comp != stk.pop():
                    return False
        return len(stk) == 0